server/scrape_cache.go

```diff
@@ -0,0 +1,322 @@
+		// or multiple string representations of the same metric. Clean up entries
+		// that haven't appeared in the last scrape.
+		for s, e := range c.series {
+			if c.iter != e.lastIter {
+				delete(c.series, s)
+			}
+		}
+		for s, iter := range c.droppedSeries {
+			if c.iter != *iter {
+				delete(c.droppedSeries, s)
+			}
+		}
+		c.metaMtx.Lock()
+		for m, e := range c.metadata {
+			// Keep metadata around for 10 scrapes after its metric disappeared.
+			if c.iter-e.lastIter > 10 {
+				delete(c.metadata, m)
+			}
+		}
+		c.metaMtx.Unlock()
+
+		c.iter++
+	}
+
+	// Swap current and previous series.
+	c.seriesPrev, c.seriesCur = c.seriesCur, c.seriesPrev
+
+	// We have to delete every single key in the map.
+	for k := range c.seriesCur {
+		delete(c.seriesCur, k)
+	}
+}
+
+func (c *scrapeCache) get(met []byte) (*cacheEntry, bool) {
+	e, ok := c.series[string(met)]
+	if !ok {
+		return nil, false
+	}
+	e.lastIter = c.iter
+	return e, true
+}
+
+func (c *scrapeCache) addRef(met []byte, ref storage.SeriesRef, lset labels.Labels, hash uint64) {
+	if ref == 0 {
+		return
+	}
+	c.series[string(met)] = &cacheEntry{ref: ref, lastIter: c.iter, lset: lset, hash: hash}
+}
+
+func (c *scrapeCache) addDropped(met []byte) {
+	iter := c.iter
+	c.droppedSeries[string(met)] = &iter
+}
+
+func (c *scrapeCache) getDropped(met []byte) bool {
+	iterp, ok := c.droppedSeries[string(met)]
+	if ok {
+		*iterp = c.iter
+	}
+	return ok
+}
+
+func (c *scrapeCache) trackStaleness(hash uint64, lset labels.Labels) {
+	c.seriesCur[hash] = lset
+}
+
+func (c *scrapeCache) forEachStale(f func(labels.Labels) bool) {
+	for h, lset := range c.seriesPrev {
+		if _, ok := c.seriesCur[h]; !ok {
+			if !f(lset) {
+				break
+			}
+		}
+	}
+}
+
+func (c *scrapeCache) setType(metric []byte, t textparse.MetricType) {
+	c.metaMtx.Lock()
+
+	e, ok := c.metadata[string(metric)]
+	if !ok {
+		e = &metaEntry{Metadata: metadata.Metadata{Type: textparse.MetricTypeUnknown}}
+		c.metadata[string(metric)] = e
+	}
+	if e.Type != t {
+		e.Type = t
+		e.lastIterChange = c.iter
+	}
+	e.lastIter = c.iter
+
+	c.metaMtx.Unlock()
+}
+
+func (c *scrapeCache) setHelp(metric, help []byte) {
+	c.metaMtx.Lock()
+
+	e, ok := c.metadata[string(metric)]
+	if !ok {
+		e = &metaEntry{Metadata: metadata.Metadata{Type: textparse.MetricTypeUnknown}}
+		c.metadata[string(metric)] = e
+	}
+	if e.Help != string(help) {
+		e.Help = string(help)
+		e.lastIterChange = c.iter
+	}
+	e.lastIter = c.iter
+
+	c.metaMtx.Unlock()
+}
+
+func (c *scrapeCache) setUnit(metric, unit []byte) {
+	c.metaMtx.Lock()
+
+	e, ok := c.metadata[string(metric)]
+	if !ok {
+		e = &metaEntry{Metadata: metadata.Metadata{Type: textparse.MetricTypeUnknown}}
+		c.metadata[string(metric)] = e
+	}
+	if e.Unit != string(unit) {
+		e.Unit = string(unit)
+		e.lastIterChange = c.iter
+	}
+	e.lastIter = c.iter
+
+	c.metaMtx.Unlock()
+}
+
+func (c *scrapeCache) GetMetadata(metric string) (scrape.MetricMetadata, bool) {
```

Comment:
We are not using `GetMetadata`, `ListMetadata`, `SizeMetadata`, `LengthMetadata` anywhere. Is it there for future use? Wondering if we can add it when actually needed? Otherwise, it's just a bit of extra unused code to review and maintain.